copy/single.go

```diff
@@ -30,6 +30,7 @@ import (
 // imageCopier tracks state specific to a single image (possibly an item of a manifest list)
 type imageCopier struct {
 	c                             *copier
+	options                       *Options
```

Comment:
`copier` already contains an `options` pointer. Why is a separate option struct passed around now?

If `imageCopier` has two sets of options, there should be a very good documentation for when to use which ones.

I _guess_ that two `options` pointers are unnecessary; if `imageCopier` should deviate from the global `options` in some way, make that _deviation_ a part of the new `singleImageOpts`.

Reply:
I’m sorry, the motivation was clearly explained in a commit message.

Yes, in that case, let’s make only the compression format difference new fields in `copySingleImageOptions`, instead of a full copy, with the intent ambiguities (and, much less important, the overhead of having to make a copy of the large `copy.Options` struct).

Reply:
Should `copySingleImageOptions` include only compression fields or entire `DestinationCtx`, for instance see `copy/single.go` its using `DestinationCtx` directly at lot of places, I'm thinking of including a custom `DestinationCtx` in `copySingleImageOptions` and then in `imageCopier` as well so other helpers can use it.

Reply:
I am not sure if `DestinationCtx` should reflect custom compression at all places in `copy/single.go` or not, I am assuming not in that case I can only add compression fields to `copySingleImageOptions`.

Reply:
Nvm I think `DestinationCtx` is not needed only compression should be enough.

Reply:
Just the compression parameters, I think. Practically, `copy/multiple.go` does not need to create a copy of `DestinationCtx` just like it shouldn’t need to create a copy of `Options`.

It _does_ make sense to reason that if `SystemContext` contains compression choices, and we are passing `SystemContext` to transports, then we should update the `SystemContext` with the chosen compression — but transports do not care (and must not care! they must write the data in PutBlob without modification for the manifest to be valid); `SystemContext.Compression…` is only used in c/image/copy, so creating an updated clone is, I think, unnecessary.

---

The history is that c/image was originally just code in a Skopeo CLI tool, where IIRC CLI parameters were just global variables. When we split c/image into a separate project, we shuffled those values to a single global `SystemContext` rather than restructure the option passing completely.

_Nowadays_, basically `SystemContext` exists primarily so that `c/image/copy` can be generic code which doesn’t know much about transports options, and callers can still set options for the transports; I’m trying to (_very slowly_) move isolated c/image subpackages away from using `SystemContext` for options, compare e.g. `signature/simplesigning`[1].  In that sense, passing the compression choice in `DestinationCtx` and not directly in `Options` is plainly not something we would do nowadays.

--- 

[1] (but also, see how much more code doing it the modern way is, compared to just adding a field to `SystemContext`)

Reply:
SGTM